LinearWeightNorm.lua

```diff
@@ -0,0 +1,220 @@
+local LinearWeightNorm, parent = torch.class('nn.LinearWeightNorm', 'nn.Linear')
+
+function LinearWeightNorm:__init(inputSize, outputSize, bias, eps)
+    nn.Module.__init(self) -- Skip nn.Linear constructor
+
+    local bias = ((bias == nil) and true) or bias
+
+    self.eps = eps or 1e-16
+
+    self.outputSize = outputSize
+    self.inputSize = inputSize
+
+    self.v = torch.Tensor(outputSize, inputSize)   
+    self.gradV = torch.Tensor(outputSize, inputSize)
+
+    self.weight = torch.Tensor(outputSize, inputSize)
+    
+    self.g = torch.Tensor(outputSize,1)
+    self.gradG = torch.Tensor(outputSize,1)
+
+    self.norm = torch.Tensor(outputSize,1)
+    self.scale = torch.Tensor(outputSize,1)
+
+    if bias then
+        self.bias = torch.Tensor(outputSize)
+        self.gradBias = torch.Tensor(outputSize)
+    end
+
+    self:reset() 
+end
+
+function LinearWeightNorm:initFromWeight(weight)
+    weight = weight or self.weight
+
+    self.g = weight:norm(2,2):add(self.eps)
+    self.v:copy(weight)
+
+    self.dirty = true
+
+    return self
+end
+
+function LinearWeightNorm.fromLinear(linear)
+    local module = nn.LinearWeightNorm(linear.weight:size(2), linear.weight:size(1), linear.bias)
+    
+    module:initFromWeight(linear.weight)
+
+    if linear.bias then
+        module.bias:copy(linear.bias)
+    end
+
+    return module
+end
+
+function LinearWeightNorm:toLinear()
+    self:updateWeightMatrix()
+
+    local module = nn.Linear(self.inputSize, self.outputSize, self.bias)
+    
+    module.weight:copy(self.weight)
+    if self.bias then
+        module.bias:copy(self.bias)
+    end
+    
+    return module
+end
+
+function LinearWeightNorm:parameters()
+    if self.bias then
+        return {self.v, self.g, self.bias}, {self.gradV, self.gradG, self.gradBias}
+    else
+        return {self.v, self.g}, {self.gradV, self.gradG}   
+    end
+end
+
+function LinearWeightNorm:reset(stdv)
+    if stdv then
+        stdv = stdv * math.sqrt(3)
+    else
+        stdv = 1 / math.sqrt(self.inputSize)
+    end
+
+    self.weight:uniform(-stdv,stdv)
+    self:initFromWeight()
+    
+    if self.bias then 
+        self.bias:uniform(-stdv, stdv)
+    end
+end
+
+local function updateAddBuffer(self, input)
+    local nframe = input:size(1)
+    self.addBuffer = self.addBuffer or input.new()
+    if self.addBuffer:nElement() ~= nframe then
+        self.addBuffer:resize(nframe):fill(1)
+    end   
+end
+
+function LinearWeightNorm:updateWeightMatrix()
+    if self.dirty or self.train then
+        if self.norm:dim() == 0 then self.norm:resizeAs(self.g) end
+        if self.scale:dim() == 0 then self.scale:resizeAs(self.g) end
+        if self.weight:dim() == 0 then self.weight:resizeAs(self.v) end
+
+        self.norm:norm(self.v,2,2):add(self.eps)
```

Comment:
It's very possible that adding this eps is what causes the unittest to fail. You could try to replace with 
```lua
self.norm:norm(self.v,2,2):clamp(self.eps, math.huge)
```